packages/gatsby-plugin-page-creator/src/babel-remove-collection-graphql.ts

```diff
@@ -0,0 +1,171 @@
+import { NodePath, PluginObj } from "@babel/core"
+import { Binding } from "@babel/traverse"
+import {
+  CallExpression,
+  Expression,
+  Identifier,
+  ImportDeclaration,
+  ObjectExpression,
+  Program,
+  StringLiteral,
+  TaggedTemplateExpression,
+} from "@babel/types"
+
+interface IGraphQLTag {
+  isGlobal: boolean
+}
+
+const isGlobalIdentifier = (
+  tag: NodePath,
+  tagName: string = `graphql`
+): boolean =>
+  tag.isIdentifier({ name: tagName }) && tag.scope.hasGlobal(tagName)
+
+export function followVariableDeclarations(binding: Binding): Binding {
+  const node = binding.path?.node
+  if (
+    node?.type === `VariableDeclarator` &&
+    node?.id.type === `Identifier` &&
+    node?.init?.type === `Identifier`
+  ) {
+    return followVariableDeclarations(
+      binding.path.scope.getBinding(node.init.name) as Binding
+    )
+  }
+  return binding
+}
+
+function getTagImport(tag: NodePath<Identifier>): NodePath | null {
+  const name = tag.node.name
+  const binding = tag.scope.getBinding(name)
+
+  if (!binding) return null
+
+  const path = binding.path
+  const parent = path.parentPath
+
+  if (
+    binding.kind === `module` &&
+    parent.isImportDeclaration() &&
+    parent.node.source.value === `gatsby`
+  )
+    return path
+
+  if (
+    path.isVariableDeclarator() &&
+    (path.get(`init`) as NodePath).isCallExpression() &&
+    (path.get(`init.callee`) as NodePath).isIdentifier({ name: `require` }) &&
+    ((path.get(`init`) as NodePath<CallExpression>).node
+      .arguments[0] as StringLiteral).value === `gatsby`
+  ) {
+    const id = path.get(`id`) as NodePath
+    if (id.isObjectPattern()) {
+      return id
+        .get(`properties`)
+        .find(
+          path => (path.get(`value`) as NodePath<Identifier>).node.name === name
+        ) as NodePath
+    }
+    return id
+  }
+  return null
+}
+
+function isGraphqlTag(tag: NodePath, tagName: string = `graphql`): boolean {
+  const isExpression = tag.isMemberExpression()
+  const identifier = isExpression ? tag.get(`object`) : tag
+
+  const importPath = getTagImport(identifier as NodePath<Identifier>)
+  if (!importPath) return isGlobalIdentifier(tag, tagName)
+
+  if (
+    isExpression &&
+    (importPath.isImportNamespaceSpecifier() || importPath.isIdentifier())
+  ) {
+    return (tag.get(`property`) as NodePath<Identifier>).node.name === tagName
+  }
+
+  if (importPath.isImportSpecifier())
+    return importPath.node.imported.name === tagName
+
+  if (importPath.isObjectProperty())
+    return (importPath.get(`key`) as NodePath<Identifier>).node.name === tagName
+
+  return false
+}
```

Comment:
I'll put my comments in a single bubble here.

- if you define the type when declaring the constant, you shouldn't need to cast as much later
- this is why we should strive to always wrap sub-statements in a block, because prettier will put it on its own line without a block if it becomes too long, which is terrible
- minor: the function could do with a comment showing some examples of what does and doesn't pass

```suggestion
function isGraphqlTag(tag: NodePath, tagName: string = `graphql`): boolean {
  const isExpression = tag.isMemberExpression()
  const identifier: NodePath<Identifier> = isExpression ? tag.get(`object`) : tag

  const importPath = getTagImport(identifier)
  if (!importPath) {
    return isGlobalIdentifier(tag, tagName)
  }

  if (
    isExpression &&
    (importPath.isImportNamespaceSpecifier() || importPath.isIdentifier())
  ) {
    return (tag.get(`property`) as NodePath<Identifier>).node.name === tagName
  }

  if (importPath.isImportSpecifier()) {
    return importPath.node.imported.name === tagName
  }

  if (importPath.isObjectProperty()) {
    return (importPath.get(`key`) as NodePath<Identifier>).node.name === tagName
  }

  return false
}
```